service/src/main/java/bio/terra/drshub/services/DrsResolutionService.java

```diff
@@ -302,19 +307,51 @@ static String getObjectId(UriComponents uriComponents) {
         Optional.ofNullable(uriComponents.getPath()).orElse(""), StandardCharsets.UTF_8);
   }
 
-  private Optional<AccessMethod> getAccessMethod(DrsObject drsResponse, DrsProvider drsProvider) {
+  private Optional<AccessMethod> getAccessMethod(DrsObject drsResponse, DrsProvider drsProvider, ResolveFromEnum resolveFrom) {
+    Optional<AccessMethod> accessMethod = Optional.empty();
+    if (!isEmpty(drsResponse)) {
+      List<AccessMethod> accessMethods = getAccessMethods(drsResponse, drsProvider);
+      if (resolveFrom != null) {
+        accessMethod = getAccessMethodForCloud(accessMethods, resolveFrom);
+      }
+      // if there is no access method matching the resolveFrom cloud, or
+      // if the resolveFrom cloud was not specified, return a different access method
+      if (accessMethod.isEmpty() & !accessMethods.isEmpty()) {
+        accessMethod = Optional.of(accessMethods.get(0));
+      }
+    }
+    return accessMethod;
+  }
+
+  private Optional<AccessMethod> getAccessMethodForCloud(List<AccessMethod> accessMethods, ResolveFromEnum resolveFrom) {
+    Optional<AccessMethod> accessMethod;
+    if (resolveFrom.equals(ResolveFromEnum.AZURE)) {
+      accessMethod =
+          accessMethods.stream()
+              .filter(m -> m.getAccessId() != null)
+              .filter(m -> m.getAccessId().startsWith("az"))
+              .findFirst();
+    } else {
+      accessMethod =
+          accessMethods.stream()
+              .filter(m -> m.getType().toString().equals(resolveFrom.toString()))
+              .findFirst();
+    }
+    return accessMethod;
+  }
```

Comment:
Since both `if` branches are the same except for the `filter()` expressions you could make the `if` branches create the filter and only filter once.
```
  private Optional<AccessMethod> getAccessMethodForCloud(List<AccessMethod> accessMethods, ResolveFromEnum resolveFrom) {
    Predicate<AccessMethod> filter;
    if (resolveFrom == ResolveFromEnum.AZURE) {
      filter = m -> m.getAccessId() != null && m.getAccessId().startsWith("az");
    } else {
      filter = m -> m.getType().toString().equals(resolveFrom.toString());
    }
    return accessMethods.stream().filter(filter).findFirst();
  }
```

Reply:
That's a good idea!